app/src/main/java/org/projectbuendia/client/ui/chart/PatientChartController.java

```diff
@@ -269,21 +269,21 @@ public void suspend() {
         }
     }
 
-    public void onXFormResult(int requestCode, int resultCode, Intent data) {
-        FormRequest request = popFormRequest(requestCode);
-        if (request == null) {
+    public void onXFormResult(final int requestCode, final int resultCode, final Intent data) {
+        final FormRequest request = popFormRequest(requestCode);
+         if (request == null) {
             LOG.e("Unknown form request code: " + requestCode);
             return;
         }
 
-        boolean shouldShowSubmissionDialog = (resultCode != Activity.RESULT_CANCELED);
-        String action = (resultCode == Activity.RESULT_CANCELED)
-            ? "form_discard_pressed" : "form_save_pressed";
-        Utils.logUserAction(action,
-            "form", request.formUuid,
-            "patient_uuid", request.patientUuid);
-        mOdkResultSender.sendOdkResultToServer(request.patientUuid, resultCode, data);
-        mUi.showFormSubmissionDialog(shouldShowSubmissionDialog);
+        final boolean isSubmissionCanceled = (resultCode == Activity.RESULT_CANCELED);
+        Utils.logUserAction(isSubmissionCanceled ? "form_discard_pressed" : "form_save_pressed",
+            "form", request.formUuid, "patient_uuid", request.patientUuid);
+        if(isSubmissionCanceled) return;
```

Comment:
@capnfabs, `OdkResultSender#sendOdkResultToServer`can still not submit the form, when it gets a validation exception. In this case, the form submition dialog would pop up indefinetely. So the second validation `isSubmittingForm = mOdkResultSender.sendOdkResultToServer(...)` is to prevent these scenario. 
If you take a look at `sendOdkResultToServer` implementation, you'll see that i returns `true` wether or not it is submitted properly, and returns false in case of validation failure. 
This behavior bothers me a bit, but i would not like to accouple `sendOdkResultToServer` with `mUi.showFormSubmissionDialog(isSubmittingForm);`, puting it inside the method. 
What do you think about that ?